File: lista9/shortestroutes2.py
def floydWarshall(adj):
    m = len(adj)
    dist = [[float('inf')] * m for _ in range(m)]
    for i in range(1,m):
        for j in range(1,m):
            if i == j:
                dist[i][j] = 0
            elif adj[i][j] != 0:
                dist[i][j] = adj[i][j]
    for k in range(1,m):
        for i in range(1,m):
            for j in range(1,m):
                if dist[i][k] + dist[k][j] < dist[i][j]:
                    dist[i][j] = dist[i][k] + dist[k][j]
    return dist

n,m,q = map(int,input().split())
adj = [[float('inf')] * (n+1) for _ in range(n+1)]
for j in range(1,n+1):
    adj[j][j] = 0
for _ in range(m):
    a,b,c = map(int,input().split())
    if c < adj[a][b]:
        adj[a][b] = c
        adj[b][a] = c
distMatrix = floydWarshall(adj)
dists = []
for _ in range(q):
    a,b = map(int,input().split())
    if distMatrix[a][b] == float('inf'): 
        dists.append(-1)
    else:
        dists.append(distMatrix[a][b])
for d in dists:
    print(d)

File: lista9/shortestroutes2.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

vector<vector<long long>> floydWarshall(vector<vector<long long>>& adj) {
    int m = adj.size();
    vector<vector<long long>> dist(m, vector<long long>(m, LLONG_MAX));

    for (int i = 1; i < m; i++) {
        for (int j = 1; j < m; j++) {
            if (i == j) {
                dist[i][j] = 0;
            } else if (adj[i][j] != LLONG_MAX) {
                dist[i][j] = adj[i][j];
            }
        }
    }

    for (int k = 1; k < m; k++) {
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < m; j++) {
                if (dist[i][k] != LLONG_MAX && dist[k][j] != LLONG_MAX &&
                    dist[i][k] + dist[k][j] < dist[i][j]) {
                    dist[i][j] = dist[i][k] + dist[k][j];
                }
            }
        }
    }

    return dist;
}

int main() {
    int n, m, q;
    cin >> n >> m >> q;

    vector<vector<long long>> adj(n + 1, vector<long long>(n + 1, LLONG_MAX));

    for (int j = 1; j <= n; ++j) {
        adj[j][j] = 0;
    }

    for (int i = 0; i < m; ++i) {
        int a, b;
        long long c;
        cin >> a >> b >> c;

        if (c < adj[a][b]) {
            adj[a][b] = c;
            adj[b][a] = c;
        }
    }

    vector<vector<long long>> distMatrix = floydWarshall(adj);

    vector<long long> dists;

    for (int i = 0; i < q; ++i) {
        int a, b;
        cin >> a >> b;

        if (distMatrix[a][b] == LLONG_MAX) {
            dists.push_back(-1);
        } else {
            dists.push_back(distMatrix[a][b]);
        }
    }

    for (const long long& d : dists) {
        cout << d << endl;
    }

    return 0;
}